hcl/raw_config.go

```diff
@@ -0,0 +1,148 @@
+// validations.
+type RawConfig struct {
+	// MergedAttributes are the top-level attributes of all files.
+	// This will be available after calling Parse or ParseConfig
+	MergedAttributes ast.Attributes
+
+	// MergedBlocks are the merged blocks from all files.
+	// This will be available after calling Parse or ParseConfig
+	MergedBlocks ast.MergedBlocks
+
+	// UnmergedBlocks are the unmerged blocks from all files.
+	// This will be available after calling Parse or ParseConfig
+	UnmergedBlocks ast.Blocks
+}
+
+// NewRawConfig returns a new RawConfig object.
+func NewRawConfig() RawConfig {
+	return RawConfig{
+		MergedAttributes: make(ast.Attributes),
+		MergedBlocks:     make(ast.MergedBlocks),
+	}
+}
+
+func (cfg *RawConfig) mergeHandlers() map[string]mergeHandler {
+	return map[string]mergeHandler{
+		"terramate":     cfg.mergeBlock,
+		"globals":       cfg.mergeBlock,
+		"stack":         cfg.addBlock,
+		"generate_file": cfg.addBlock,
+		"generate_hcl":  cfg.addBlock,
+		"import":        func(b *ast.Block) error { return nil },
+	}
+}
+
+// Merge the config with the provided other config.
+func (cfg *RawConfig) Merge(other RawConfig) error {
+	errs := errors.L()
+	errs.Append(cfg.mergeAttrs(other.MergedAttributes))
+	errs.Append(cfg.mergeBlocks(other.MergedBlocks.AsBlocks()))
+	errs.Append(cfg.mergeBlocks(other.UnmergedBlocks))
+	return errs.AsError()
+}
+
+func (cfg *RawConfig) mergeBlocks(blocks ast.Blocks) error {
+	handlers := cfg.mergeHandlers()
+
+	errs := errors.L()
+	for _, block := range blocks {
+		handler, ok := handlers[block.Type]
+		if !ok {
+			errs.Append(
+				errors.E(ErrTerramateSchema, block.DefRange(),
+					"unrecognized block %q", block.Type),
+			)
+
+			continue
+		}
+
+		errs.Append(handler(block))
+	}
+	return errs.AsError()
+}
+
+func (cfg *RawConfig) addBlock(block *ast.Block) error {
+	cfg.UnmergedBlocks = append(cfg.UnmergedBlocks, block)
+	return nil
+}
+
+func (cfg *RawConfig) mergeBlock(block *ast.Block) error {
+	if other, ok := cfg.MergedBlocks[block.Type]; ok {
+		err := other.MergeBlock(block)
+		if err != nil {
+			return errors.E(ErrTerramateSchema, err)
+		}
+		return nil
+	}
+
+	merged := ast.NewMergedBlock(block.Type)
+	cfg.MergedBlocks[block.Type] = merged
+	err := merged.MergeBlock(block)
+	if err != nil {
+		return errors.E(ErrTerramateSchema, err)
+	}
+	return nil
+}
+
+func (cfg *RawConfig) mergeAttrs(other ast.Attributes) error {
+	errs := errors.L()
+	for _, attr := range other.SortedList() {
+		if attrVal, ok := cfg.MergedAttributes[attr.Name]; ok &&
+			sameDir(attrVal.Origin, attr.Origin) {
+			errs.Append(errors.E(ErrTerramateSchema,
+				attr.NameRange,
+				"attribute %q redeclared", attr.Name))
+			continue
+		}
+
+		cfg.MergedAttributes[attr.Name] = attr
+	}
+	return errs.AsError()
+}
+
+func (cfg RawConfig) filterUnmergedBlocksByType(blocktype string) ast.Blocks {
+	logger := log.With().
+		Str("action", "filterUnmergedBlocksByType()").
```

Comment:
```suggestion
		Str("action", "RawConfig.filterUnmergedBlocksByType()").
```

I know we are not consistent on this today, but we need to start having some more namespacing on the action field or it may be hard to map the log entries to actual code etc (Im prone to object type for methods and pkg name for functions).